notebooks/community/aisamples/AIsample - Book Recommendation.ipynb

```diff
@@ -1254,9 +1259,60 @@
    },
    "outputs": [],
    "source": [
-    "saved_model_path = f\"{DATA_FOLDER}/als_model/{int(time.time())}\"\n",
-    "model.save(saved_model_path)\n",
-    "print(f\"model is saved in: {saved_model_path}\")"
+    "# setup mlflow\n",
+    "import mlflow\n",
+    "import trident.mlflow\n",
+    "from trident.mlflow import get_sds_url\n",
+    "\n",
+    "mlflow.set_tracking_uri(get_sds_url())\n",
+    "mlflow.set_registry_uri(get_sds_url())\n",
+    "mlflow.set_experiment(EXPERIMENT_NAME)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# log model, metrics and params\n",
+    "with mlflow.start_run() as run:\n",
+    "    print(\"log model:\")\n",
+    "    mlflow.spark.log_model(\n",
+    "        model,\n",
+    "        f\"{EXPERIMENT_NAME}-alsmodel\",\n",
+    "        registered_model_name=f\"{EXPERIMENT_NAME}-alsmodel\",\n",
+    "        dfs_tmpdir=\"Files/spark\",\n",
+    "    )\n",
+    "\n",
+    "    print(\"log metrics:\")\n",
+    "    mlflow.log_metrics({\"RMSE\": rmse, \"MAE\": mae, \"R2\": r2, \"Explained variance\": var})\n",
+    "\n",
+    "    print(\"log parameters:\")\n",
+    "    mlflow.log_params(\n",
+    "        {\n",
+    "            \"num_epochs\": num_epochs,\n",
+    "            \"rank_size_list\": rank_size_list,\n",
+    "            \"reg_param_list\": reg_param_list,\n",
+    "            \"model_tuning_method\": model_tuning_method,\n",
+    "            \"DATA_FOLDER\": DATA_FOLDER,\n",
+    "        }\n",
+    "    )\n",
+    "\n",
+    "    model_uri = f\"runs:/{run.info.run_id}/{EXPERIMENT_NAME}-alsmodel\"\n",
+    "    print(\"Model saved in run %s\" % run.info.run_id)\n",
+    "    print(f\"Model URI: {model_uri}\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# load model back\n",
+    "# mlflow will use PipelineModel to wrapper the original model, thus here we extract the original ALSModel from the stages.\n",
+    "loaded_model = mlflow.spark.load_model(model_uri, dfs_tmpdir=\"Files/spark\").stages[-1]"
```

Comment:
I think serena might have fixed this behavior in MLFlow so that it just loads and saves original model, can you check with her to see?

Reply:
Hihi, I didn't change this part lol and we can't. Because they need a consistent API to load the model back, which only PipelineModel does, so we have to fetch the stages within it. I was using the similar logic in our dotnet stuff, because you can't know the real type only after you load it back, and sometimes we might just save a PipelineModel so you don't need to fetch the stages.